src/HCSR04.cpp

```diff
@@ -20,21 +22,30 @@ double UltraSonicDistanceSensor::measureDistanceCm() {
 }
 
 double UltraSonicDistanceSensor::measureDistanceCm(float temperature) {
+    uint32_t maxDistanceDurationMicroSec;
+
     // Make sure that trigger pin is LOW.
     digitalWrite(triggerPin, LOW);
     delayMicroseconds(2);
     // Hold trigger for 10 microseconds, which is signal for sensor to measure distance.
     digitalWrite(triggerPin, HIGH);
     delayMicroseconds(10);
     digitalWrite(triggerPin, LOW);
+    double speedOfSoundInCmPerMicroSec = 0.03313 + 0.0000606 * temperature; // Cair ≈ (331.3 + 0.606 ⋅ ϑ) m/s
+
+    // Compute max delay based on max distance with 25% margin in microseconds or use absolute timeout
+    maxDistanceDurationMicroSec = 2.5 * maxDistanceCm / speedOfSoundInCmPerMicroSec;
+    if (maxTimeoutMicroSec > 0) {
+    	maxDistanceDurationMicroSec = min(maxDistanceDurationMicroSec, maxTimeoutMicroSec);
+    }
+
     // Measure the length of echo signal, which is equal to the time needed for sound to go there and back.
-    unsigned long durationMicroSec = pulseIn(echoPin, HIGH);
+    uint32_t durationMicroSec = pulseIn(echoPin, HIGH, maxDistanceDurationMicroSec); // can't measure beyond max distance
```

Comment:
I did short investigation in uint32_t and I am slightly concerned with using it and unsigned long interchangeably. This works because it seems that on arduino unsigned long actually is uint32_t, but if we don't need to be making this assumption, why do it? Why not just use unsigned long all the way and we are good?

Reply:
I think you're right, it's more convenient to use the same type as as pulseIn.